node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage> + Sync + Send,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		SpawnedSubsystem {
+			name: "availability-distribution",
+			future: Box::pin(async move { self.run(ctx) }.map(|_| ())),
+		}
+	}
+}
+
+/// Obtain all live candidates based on an iterator of relay heads.
+async fn query_live_candidates_without_ancestors<Context>(
+	ctx: &mut Context,
+	relay_parents: impl IntoIterator<Item = H256>,
+) -> Result<HashSet<CommittedCandidateReceipt>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let iter = relay_parents.into_iter();
+	let hint = iter.size_hint();
+
+	let mut live_candidates = HashSet::with_capacity(hint.1.unwrap_or(hint.0));
+	for relay_parent in iter {
+		let paras = query_para_ids(ctx, relay_parent.clone()).await?;
+		for para in paras {
+			if let Some(ccr) = query_pending_availability(ctx, relay_parent, para).await? {
+				live_candidates.insert(ccr);
+			}
+		}
+	}
+	Ok(live_candidates)
+}
+
+/// Obtain all live canddidates based on an iterator or relay heads including `k` ancestors.
+///
+/// Relay parent.
+async fn query_live_candidates<Context>(
+	ctx: &mut Context,
+	state: &mut ProtocolState,
+	relay_parents: impl IntoIterator<Item = H256>,
+) -> Result<HashMap<H256, CommittedCandidateReceipt>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let iter = relay_parents.into_iter();
+	let hint = iter.size_hint();
+
+	let capacity = hint.1.unwrap_or(hint.0) * (1 + AvailabilityDistributionSubsystem::K);
+	let mut live_candidates = HashMap::<H256, CommittedCandidateReceipt>::with_capacity(capacity);
+
+	for relay_parent in iter {
+		// direct candidates for one relay parent
+		let receipts =
+			query_live_candidates_without_ancestors(ctx, iter::once(relay_parent.clone())).await?;
+		live_candidates.extend(iter::once(relay_parent.clone()).zip(receipts.into_iter()));
+
+		// register one of relay parents (not the ancestors)
+		let ancestors =
+			query_k_ancestors(ctx, relay_parent, AvailabilityDistributionSubsystem::K).await?;
+
+		// ancestors might overlap, so check t	he cache too
+		let unknown = ancestors
+			.into_iter()
+			.filter(|relay_parent| {
+				// use the ones which we pulled before
+				// but keep the unknown relay parents
+				state
+					.receipts
+					.get(relay_parent)
+					.and_then(|receipts| {
+						// directly extend the live_candidates with the cached value
+						live_candidates.extend(
+							receipts
+								.into_iter()
+								.map(|receipt| (relay_parent.clone(), receipt.clone())),
+						);
+						Some(())
+					})
+					.is_none()
+			})
+			.collect::<Vec<_>>();
+
+		// query the ones that were not present in the receipts cache
+		let receipts = query_live_candidates_without_ancestors(ctx, unknown.clone()).await?;
+		live_candidates.extend(unknown.into_iter().zip(receipts.into_iter()));
```

Comment:
should we write into the receipts cache here?

Reply:
That happens at the call site, but it's better to move it into the this `fn` :+1: